### PR TITLE
Realize02 03 2019

### PR DESCRIPTION
**Pull request by Sofasmile**

### CODE REVIEW

File: src/com/collection/MyArrayList.java [hunk 1]
```diff
@@ -1,4 +1,70 @@
...
+
+    @Override
+    public void add(T value) {
+        if (size == arrays.length) {
+            increazeSize();
+        }
+        arrays[size++] = value;
+    }
+
+    private void increazeSize() {
```
Comment (Anderwerty) [inline src/com/collection/MyArrayList.java]: Typo. Should be increaseSize

File: src/com/collection/MyArrayList.java [hunk 2]
```diff
@@ -1,4 +1,70 @@
...
+    @Override
+    public int size() {
+        return size;
+    }
+
+    @Override
+    public T get(int i) {
+        if (i < 0 || i >= size) {
+            System.out.println("Index " + i + " is out of bound!");
+            return null;
```
Comment (Anderwerty) [inline src/com/collection/MyArrayList.java]: It is not a very good solution.  As one or more of elements in the list could be null. It is better to throw the custom exception.

File: src/com/collection/MyLinkedList.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+            node.prev = null;
+        }
+        node.data = null;
+        size--;
+
+    }
+
+    private Node<T> getNode(int index) {
+        if (index < 0 || index >= size) {
+            System.out.println("Index " + index + " is out of bound!");
```
Comment (Anderwerty) [inline src/com/collection/MyLinkedList.java]: The same

File: src/com/collection/MyLinkedList.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            for (int i = size - 1; i > index; i--) {
+                item = item.prev;
+            }
+            return item;
+        }
+    }
+
+    @Override
+    public void clear() {
+        int oldSize = size;
```
Comment (Anderwerty) [inline src/com/collection/MyLinkedList.java:84]: Do we really need variable `oldSize`?

File: src/com/collection/MyLinkedList.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+    @Override
+    public T get(int i) {
+        return getNode(i).data;
+    }
+
+    public void show() {
+        Node<T> iterator = first;
+        for (int i = 0; i < size; i++) {
+            if (iterator == null) {
```
Comment (Anderwerty) [inline src/com/collection/MyLinkedList.java]: Is it possible situation when the iterator is null and size>0?

File: src/com/collection/MyStack.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
+package collection;
+
+public class MyStack<T> {
+    private NodeStack<T> head;
+    int size;
```
Comment (Anderwerty) [inline src/com/collection/MyStack.java]: private?

File: src/com/collection/MyStack.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        } else {
+            prev.next = node.next;
+        }
+        size--;
+    }
+
+    public void show() {
+        NodeStack<T> item = head;
+        for (int i = 0; i < size; i++) {
+            if (item == null) {
```
Comment (Anderwerty) [inline src/com/collection/MyStack.java]: The same

**Review by Anderwerty — CHANGES_REQUESTED** [review]
Fix comments. And look at your .gitignore.

File: src/com/collection/MyLinkedList.java (hunk 4 above)
Comment (Sofasmile) [reply src/com/collection/MyLinkedList.java:84]: Так, потрібна) Адже при видаленні, size  зменшується на 1, після кожного виклику методу remove(0).

File: src/com/collection/Node.java [hunk 8]
```diff
@@ -0,0 +1,13 @@
+package collection;
+
+public class Node<T> {
+    public Node<T> prev;
```
Comment (Anderwerty) [inline src/com/collection/Node.java]: Why not private? You could use Node as a static inner class in MyLinkedList.

File: src/com/collection/MyLinkedList.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    private Node<T> last;
+    private int size;
+
+    public MyLinkedList() {
+        first = new Node<>(null, null, null);
+        last = first;
+        size = 0;
+    }
+
+    //addLast
```
Comment (Anderwerty) [inline src/com/collection/MyLinkedList.java]: Do you need this comment?

File: src/com/collection/MyLinkedList.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    @Override
+    public void add(T item) {
+        Node<T> newNode = new Node<>(last, item, null);
+        if (size == 0) {
+            first = newNode;
+        }
+        last.next = newNode;
```
Comment (Anderwerty) [inline src/com/collection/MyLinkedList.java:32]: It looks like a singly linked list.
Please, read https://habr.com/ru/post/127864/
You don't use Node.prev

File: src/com/collection/MyLinkedList.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+            node.prev = null;
+        }
+        node.data = null;
+        size--;
+
+    }
+
+    private Node<T> getNode(int index) throws Exception {
+        if (index < 0 || index >= size) {
+            throw new Exception("Index is out of bound!");
```
Comment (Anderwerty) [inline src/com/collection/MyLinkedList.java]: Please. Create your own Exception. For example NodeIndexOutBoundaryException (extends RuntimeException).

File: src/com/collection/MyStack.java [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+    public void push(T item) {
+        NodeStack<T> node = new NodeStack<>(item, head);
+        head = node;
+        size++;
+    }
+
+    public T pop() {
+        if (isEmpty()) {
+            System.out.println("Steck is empty!");
```
Comment (Anderwerty) [inline src/com/collection/MyStack.java]: Throw your own exception.

File: src/com/collection/MyStack.java [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+        }
+        NodeStack<T> temp = head;
+        head = head.next;
+        size--;
+        return temp.data;
+    }
+
+    public T peek() {
+        if (isEmpty()) {
+            System.out.println("Steck is empty!");
```
Comment (Anderwerty) [inline src/com/collection/MyStack.java]: Throw your own exception.

File: src/com/collection/MyStack.java [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+    public void clear() {
+        int oldSize = size;
+        for (int i = 0; i < oldSize; i++) {
+            pop();
+        }
+    }
+
+    public void remove(int index) {
+        if (isEmpty()) {
+            System.out.println("Steck is empty!");
```
Comment (Anderwerty) [inline src/com/collection/MyStack.java]: Throw your own exception.

File: src/com/collection/MyStack.java [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+            return;
+        }
+        if (index >= size || index < 0) {
+            System.out.println("Index " + index + " is out of bound!");
```
Comment (Anderwerty) [inline src/com/collection/MyStack.java]: Throw your own exception.

**Review by Anderwerty — CHANGES_REQUESTED** [review]
Fix comments

File: src/com/collection/MyLinkedList.java (hunk 11 above)
Comment (Sofasmile) [reply src/com/collection/MyLinkedList.java]: На жаль, не впевнена що правильно створила свої класи exception.

File: src/com/collection/MyLinkedList.java [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+            next.prev = prev;
+            node.next = null;
+            node.prev = null;
+        }
+        node.data = null;
+        size--;
+
+    }
+
+    private Node<T> getNode(int index) throws NodeIndexOutBoundaryException {
```
Comment (Anderwerty) [inline src/com/collection/MyLinkedList.java]: If NodeIndexOutBoundaryException extends from RuntimeException you don't need to write throws NodeIndexOutBoundaryException. It is an unchecked exception.

**Review by Anderwerty — APPROVED** [review]
Fix small commetn

**Pull request merged**